src/main/java/org/spongepowered/api/data/DataTransactionBuilder.java

```diff
@@ -0,0 +1,309 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.data;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import org.spongepowered.api.data.value.BaseValue;
+import org.spongepowered.api.data.value.immutable.ImmutableValue;
+import org.spongepowered.api.data.value.mutable.Value;
+import org.spongepowered.api.event.data.ValueChangeEvent;
+
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * A type of builder for building {@link DataTransactionResult}s. The common
+ * use is for both implementations of {@link DataHolder}s, and various
+ * {@link ValueChangeEvent}s.
+ */
+public final class DataTransactionBuilder {
+
+    @Nullable private List<ImmutableValue<?>> rejected;
+    @Nullable private List<ImmutableValue<?>> replaced;
+    @Nullable private List<ImmutableValue<?>> successful;
+    private DataTransactionResult.Type resultType;
+
+    private DataTransactionBuilder() {
+    }
+
+    /**
+     * Gets a new {@link DataTransactionBuilder} to build a new
+     * {@link DataTransactionResult}.
+     *
+     * @return The new builder, for chaining
+     */
+    public static DataTransactionBuilder builder() {
+        return new DataTransactionBuilder();
+    }
+
+    /**
+     * Creates a {@link DataTransactionResult} with no data successfully added,
+     * removed, or rejected, and with the
+     * {@link org.spongepowered.api.data.DataTransactionResult.Type} of
+     * {@link org.spongepowered.api.data.DataTransactionResult.Type#SUCCESS}
```

Comment:
> Fully qualifying makes this more difficult to read.

Yes, but absolutely necessary since one of the JDK's doesn't locate the linked enum value correctly.